File: 235/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        value = root.val
        left_value = p.val
        right_value = q.val

        if left_value > value and right_value > value:
            return self.lowestCommonAncestor(root.right, p, q)
        elif left_value < value and right_value < value:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root

File: 235/LowestCommonAncestorOfABinarySearchTree.java
/**
 * LowestCommonAncestorOfABinarySearchTree
 */
public class LowestCommonAncestorOfABinarySearchTree {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int value, leftValue = p.val, rightValue = q.val;
        
        while (null != root) {
            value = root.val;
            if (leftValue > value && rightValue > value) {
                root = root.right;
            } else if (leftValue < value && rightValue < value) {
                root = root.left;
            } else break;
        }
        return root;
    }
    
}